tests/e2e/test_spyre_cb.py

```diff
@@ -83,177 +91,412 @@ def test_cb_handling(
+
+
+def create_random_request(
+        request_id: int, num_tokens: int,
+        sampling_params: SamplingParams) -> EngineCoreRequest:
+    return EngineCoreRequest(
+        request_id=str(request_id),
+        prompt_token_ids=[request_id] * num_tokens,
+        mm_inputs=None,
+        mm_hashes=None,
+        mm_placeholders=None,
+        sampling_params=sampling_params,
+        eos_token_id=None,
+        arrival_time=0,
+        lora_request=None,
+    )
+
+
+def get_params_test_blocks_borders_aligned_prompts():
+    seqs_max_tokens = [65, 67, 7]
+    prompts_lengths = [49, 41, 7]
+    steps_add_reqs = [0, 0, 0]  # add all requests in the beginning
+
+    checked_steps = [{
+        "step": 0,
+        "tkv": 0,
+        "num_waiting": 3,
+        "num_running": 0,
+        "request_outputs": []
+    }, {
+        "step": 1,  # Prefill sequence 0
+        "tkv": 64,
+        "num_waiting": 2,
+        "num_running": 1,
+        "request_outputs": ["0"]  
+    }, {
+        "step": 2,  # Prefill sequence 1
+        "tkv": 64,  # Still 64 because this step is also a prefill
+        "num_waiting": 1,
+        "num_running": 2,
+        "request_outputs": ["1"]  
+    }, {
+        "step": 3,
+        "tkv": 65,  # Two decodes increases the tkv
+        "num_waiting": 1,
+        "num_running": 2,
+        "request_outputs": ["1", "0"]  # Two sequences are decoded
+    }, {
+        "step": 4,  # Check normal decode continuation
+        "tkv": 66,
+        "num_waiting": 1,
+        "num_running": 2,
+        "request_outputs": ["1", "0"]
+    }, {
+        "step": 65,  # Last step before fist sequence finishes
+        "tkv": 127,
+        "num_waiting": 1,
+        "num_running": 2,
+        "request_outputs": ["1", "0"]
+    # }, {
+    #     "step": 66,  # Sequence 0 finishes at step 66 (2 prefills + 64 decodes)
+    #     "tkv": 128,
+    #     "num_waiting": 1,
+    #     "num_running": 1,
+    #     "request_outputs": ["1"]  # TODO this fails?? Why do we observe ["1", "0"]?
+    }, {
+        "step": 67,  # Prefill sequence 2
+        "tkv": 128,  # Tkv doesn't increase because it is a prefill
+        "num_waiting": 0,
+        "num_running": 2,
+        "request_outputs": ["2"]
+    }, {
+        "step": 68,  # Decode sequences 1 and 2
+        "tkv": 129,
+        "num_waiting": 0,
+        "num_running": 2,
+        "request_outputs": ["2", "1"]
+    }, {
+        # Sequence 1 finishes at step 71 (start step + 2 prefills + 67 decodes)
+        "step": 71,
+        "tkv": 132,
+        "num_waiting": 0,
+        "num_running": 1,
+        "request_outputs": ["2"]
+    }, {
+        "step": 72,  # Decode sequence 2
+        "tkv": 133,
+        "num_waiting": 0,
+        "num_running": 1,
+        "request_outputs": ["2"]
+    }, {
+        # Sequence 2 finishes at step 73 (start step + 1 prefill + 7 decodes)
+        "step": 73,  
+        "tkv": 134,
+        "num_waiting": 0,
+        "num_running": 1,
```

Comment:
This step wasn't running since the loop only looped till 73 (excluding it). I have a suggestion below on how to fix it instead of using `last_step` explicitly
```suggestion
        "num_running": 0,
```

Reply:
You're right, I'll change that